onnxruntime/contrib_ops/cpu/triu.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/platform/threadpool.h"
+#include "core/util/math_cpuonly.h"
+#include "Eigen/src/Core/Map.h"
+#include "Eigen/LU"
+#include "triu.h"
+#include <functional>
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_TYPED_KERNEL_EX(
+    Triu,
+    kMSDomain,
+    1,
+    float,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+    Triu);
+
+ONNX_OPERATOR_TYPED_KERNEL_EX(
+    Triu,
+    kMSDomain,
+    1,
+    double,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::GetTensorType<double>()),
+    Triu);
+
+template <typename T>
+struct Triu::ComputeImpl {
+
+  void get_triu(const T* X_data, T* Y_data, int64_t matrix_h, int64_t matrix_w, int64_t k_val) const {
+    auto input_mat = ConstEigenMatrixMapRowMajor<T>(
+	      X_data,
+	      matrix_h,
+	      matrix_w);
+	  auto output_mat = EigenMatrixMapRowMajor<T>(
+	      Y_data,
+	      matrix_h,
+	      matrix_w);
+
+	  output_mat = input_mat;
+	  for (int64_t i = -1 * matrix_h; i < k_val; i++){
+	    output_mat.diagonal(i).array() = static_cast<T>(0);
+	  }
+  }
+
+  void operator()(const Tensor* X, const Tensor* k, Tensor* Y) const {
+	  int64_t k_val = 0;
+	  if (k) {
+	    ORT_ENFORCE(k->Shape().NumDimensions() <= 1, "k should be a 1-D or 0-D tensor.");
```

Comment:
Maybe use IsScalarOr1ElementVector from core\providers\common.h? This also verifies that a 1D element has exactly one element.

Reply:
The generated code would also be smaller if this were lifted out to the Compute() function instead of inside the ComputeImpl. The function could be changed to pass in k_val instead of the K tensor.